### ISSUE TITLE
Practically non-functional behind a CGNAT internet connection

### ISSUE THREAD

**Issue by Techjar** [Spore-Community/ModBrowser#12]
This tool is using client-side unauthenticated requests to the GitHub API, which have a rate limit of 60 requests per hour per IP address. CGNAT shares an IP address with thousands of users doing who knows what, so chances are the rate limit will be exhausted very quickly every time it resets, rendering this tool essentially unusable.

~~The only solution is, unfortunately, to move the API requests to the server side and use GitHub's OAuth2 flow to obtain a token which can be sent in requests to the API. You'll likely also need to implement some sort of server-side caching to avoid hitting the authenticated rate limit.~~ Re-reading the documentation, I think you can allow the user to "sign in" with their GitHub account, and grant the application an OAuth2 token unique to them which can be held in local storage. This way, the requests can remain client-side and are tied to each user rather than the whole server.

Here's a link to the relevant documentation: https://docs.github.com/en/rest/overview/resources-in-the-rest-api#rate-limiting

**Comment by Splitwirez** [comment]
This "60 requests per hour per IP address" rate limit...do you know at what scope it applies?
With respect to a given repo?
With respect to all repos owned by a given GitHub user?
With respect to..the whole of GitHub...?

**Comment by Techjar** [comment]
Ah sorry, I've added a link to the documentation. As I understand it applies globally to any kind of request, only discriminating by IP address, per this paragraph:

> For unauthenticated requests, the rate limit allows for up to 60 requests per hour. Unauthenticated requests are associated with the originating IP address, and not the person making requests.

So the rate limit is likely being exhausted by various other apps that make client-side API requests in the same way as this one (I doubt that many people behind the same IP are using the Spore mod browser). I also updated the second paragraph of the top post after going over the docs again.
